### PR TITLE
Raise warning message for insecure protocols

### PR DESCRIPTION
**Pull request by DavoudEshtehari**
For protocols, lower than Tls 1.2 raises a warning message by `Console.Out` in addition to `EventSource` at `Netfx` and `NetCore`.

### CODE REVIEW

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.Windows.cs [hunk 1]
```diff
@@ -75,15 +75,13 @@ private void LoadSSPILibrary()
             }
         }
 
-        private void WaitForSSLHandShakeToComplete(ref uint error)
+        private void WaitForSSLHandShakeToComplete(ref uint error, ref int protocolVersion)
         {
-            if (TdsParserStateObjectFactory.UseManagedSNI)
-                return;
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.Windows.cs:81]: Is there no possibility of control coming here for Managed SNI?

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserHelperClasses.cs [hunk 2]
```diff
@@ -917,4 +918,58 @@ private void ParseMultipartName()
...
+            {
+                name = "TLS 1.0";
+            }
+#pragma warning disable CS0618 // Type or member is obsolete: SSL is depricated
+            else if ((protocol & SslProtocols.Ssl3) == SslProtocols.Ssl3)
+            {
+                name = "SSL 3.0";
+            }
+            else if ((protocol & SslProtocols.Ssl2) == SslProtocols.Ssl2)
+#pragma warning disable CS0618 // Type or member is obsolete: SSL is depricated
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserHelperClasses.cs:952]: You aren't restoring this warning later, please do so here too.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserHelperClasses.cs [hunk 3]
```diff
@@ -917,4 +918,58 @@ private void ParseMultipartName()
...
+                name = "SSL 2.0";
+            }
+            else
+            {
+                name = protocol.ToString();
+            }
+
+            return name;
+        }
+        public static string GetProtocolWarning(this SslProtocols protocol)
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserHelperClasses.cs:963]: Nit: Please add extra line before method.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObjectNative.cs [hunk 4]
```diff
@@ -6,13 +6,33 @@
...
 using System.Runtime.InteropServices;
+using System.Security.Authentication;
 using System.Threading.Tasks;
 using Microsoft.Data.Common;
 
 namespace Microsoft.Data.SqlClient
 {
     internal class TdsParserStateObjectNative : TdsParserStateObject
     {
+        // ptotocol versions from native sni
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObjectNative.cs:17]: Nit: 
```suggestion
        // protocol versions from native sni
```

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObjectNative.cs [hunk 5]
```diff
@@ -309,8 +330,49 @@ internal override uint SetConnectionBufferSize(ref uint unsignedPacketSize)
         internal override uint GenerateSspiClientContext(byte[] receivedBuff, uint receivedLength, ref byte[] sendBuff, ref uint sendLength, byte[] _sniSpnBuffer)
             => SNINativeMethodWrapper.SNISecGenClientContext(Handle, receivedBuff, receivedLength, sendBuff, ref sendLength, _sniSpnBuffer);
 
-        internal override uint WaitForSSLHandShakeToComplete()
-            => SNINativeMethodWrapper.SNIWaitForSSLHandshakeToComplete(Handle, GetTimeoutRemaining());
+        internal override uint WaitForSSLHandShakeToComplete(out int protocolVersion)
+        {
+            uint returnValue = SNINativeMethodWrapper.SNIWaitForSSLHandshakeToComplete(Handle, GetTimeoutRemaining(), out uint nativeProtocolVersion);
+            var nativeProtocol = (NativeProtocols)nativeProtocolVersion;
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObjectNative.cs:336]: Could this be null by any chance?
Consider a null check for `nativeProtocolVersion`

File: src/Microsoft.Data.SqlClient/netcore/src/Resources/SR.resx [hunk 6]
```diff
@@ -1863,4 +1863,7 @@
   <data name="SQL_BulkLoadCannotConvertValueWithoutRowNo" xml:space="preserve">
     <value>The given value{0} of type {1} from the data source cannot be converted to type {2} for Column {3} [{4}].</value>
   </data>
-</root>
+  <data name="SEC_ProtocolWarning" xml:space="preserve">
+    <value>Security Warning: The {0} is an unsecured protocol and is provided for backward compatibility only. The recommended protocol is TLS 1.2 and later.</value>
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netcore/src/Resources/SR.resx:1867]:
```suggestion
    <value>Security Warning: The negotiated {0} is an insecure protocol and is supported for backward compatibility only. The recommended protocol version is TLS 1.2 and later.</value>
```

File: src/Microsoft.Data.SqlClient/netcore/src/Resources/SR.resx [hunk 7]
```diff
@@ -1863,4 +1863,7 @@
...
+  </data>
+</root>
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netcore/src/Resources/SR.resx:1869]: Nit: EOF line

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/TdsParserHelperClasses.cs [hunk 8]
```diff
@@ -1421,4 +1421,82 @@ private void ParseMultipartName()
 
         internal static readonly MultiPartTableName Null = new MultiPartTableName(new string[] { null, null, null, null });
     }
+
+    internal static class SslProtocolsHelper
+    {
+        // ptotocol versions from native sni
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/TdsParserHelperClasses.cs:1427]:
```suggestion
        // protocol versions from native sni
```

File: src/Microsoft.Data.SqlClient/netfx/src/Resources/Strings.resx [hunk 9]
```diff
@@ -4530,4 +4530,10 @@
   <data name="SQLUDT_InvalidSize" xml:space="preserve">
     <value>UDT size must be less than {1}, size: {0}</value>
   </data>
+  <data name="SEC_ProtocolWarning" xml:space="preserve">
+    <value>Security Warning: The {0} is an unsecured protocol and is provided for backward compatibility only. The recommended protocol is TLS 1.2 and later.</value>
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netfx/src/Resources/Strings.resx:4534]:
```suggestion
    <value>Security Warning: The negotiated {0} is an insecure protocol and is supported for backward compatibility only. The recommended protocol version is TLS 1.2 and later.</value>
```

File: src/Microsoft.Data.SqlClient/netfx/src/Resources/Strings.resx [hunk 10]
```diff
@@ -4530,4 +4530,10 @@
...
+  </data>
+  <data name="net_invalid_enum" xml:space="preserve">
+    <value>The specified value is not valid in the '{0}' enumeration.</value>
+  </data>
 </root>
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netfx/src/Resources/Strings.resx:4539]: Nit: EOF Line

File: src/NuGet.config [hunk 11]
```diff
@@ -5,5 +5,6 @@
     <add key="dotnet-core" value="https://dotnetfeed.blob.core.windows.net/dotnet-core/index.json" />
     <add key="nuget.org" value="https://api.nuget.org/v3/index.json" />
     <add key="dotnet-eng" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-eng/nuget/v3/index.json" />
+    <add key="Package source" value="C:\Packages" />
```
Comment (cheenamalhotra) [inline src/NuGet.config:8]: Accidental commit :)

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.Windows.cs (hunk 1 above)
Comment (DavoudEshtehari) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.Windows.cs:81]: This part is managed by overriding the `TdsParserStateObjectManaged.WaitForSSLHandShakeToComplete()`.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObjectNative.cs (hunk 5 above)
Comment (DavoudEshtehari) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObjectNative.cs:336]: I created this object in TdsParser and assigned a default value to it. But, I think it is a good practice to do a null check.
Thank you.

Comment (DavoudEshtehari) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObjectNative.cs:336]: In manged part no. But for safety, I will check it.
Thank you

File: tools/specs/Microsoft.Data.SqlClient.nuspec [hunk 12]
```diff
@@ -9,19 +9,21 @@
...
-Microsoft.Data.SqlClient.SqlException
-Microsoft.Data.SqlClient.SqlParameter
-Microsoft.Data.SqlClient.SqlDataReader
-Microsoft.Data.SqlClient.SqlCommand
-Microsoft.Data.SqlClient.SqlTransaction
-Microsoft.Data.SqlClient.SqlParameterCollection
-Microsoft.Data.SqlClient.SqlClientFactory
-
-When using NuGet 3.x this package requires at least version 3.4.</description>
+    <description>
```
Comment (karinazhou) [inline tools/specs/Microsoft.Data.SqlClient.nuspec:12]: Please revert the change in this file.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObjectNative.cs [hunk 13]
```diff
@@ -309,8 +330,49 @@ internal override uint SetConnectionBufferSize(ref uint unsignedPacketSize)
...
+                protocolVersion = (int)SslProtocols.Ssl3;
+            }
+            else if (nativeProtocol.HasFlag(NativeProtocols.SP_PROT_SSL2_CLIENT) || nativeProtocol.HasFlag(NativeProtocols.SP_PROT_SSL2_SERVER))
+            {
+                protocolVersion = (int)SslProtocols.Ssl2;
+#pragma warning restore CS0618 // Type or member is obsolete : SSL is depricated
+            }
+            else if(nativeProtocol.HasFlag(NativeProtocols.SP_PROT_NONE))
+            {
+                protocolVersion = (int)SslProtocols.None;
```
Comment (karinazhou) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObjectNative.cs:368]: Will this trigger warning CS0618 since SSL is deprecated?

And line 364 will not be reached when we hit line 358.

Comment (DavoudEshtehari) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObjectNative.cs:368]: Yes, this warning triggers for obsolete members. You can try it by removing this line.
[#pragma warning](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/preprocessor-directives/preprocessor-pragma-warning) is a preprocessor directive and you can use it everywhere in the code.

Comment (karinazhou) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObjectNative.cs:368]: > And line 364 will not be reached when we hit line 358.

Yeah, I just realize the preprocessor won't be impacted by common coding logic. Just ignore this comment. 

> Will this trigger warning CS0618 since SSL is deprecated?

This is for `protocolVersion = (int)SslProtocols.None;` which is not covered by `#pragma warning disable CS0618 `.

Comment (DavoudEshtehari) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObjectNative.cs:368]: The `SslProtocols.None` does not trigger warning CS0618. It just lets the operating system to choose the best protocol by itself when you want to establish a secure connection([MS document](https://docs.microsoft.com/en-us/dotnet/api/system.security.authentication.sslprotocols?f1url=https%3A%2F%2Fmsdn.microsoft.com%2Fquery%2Fdev16.query%3FappId%3DDev16IDEF1%26l%3DEN-US%26k%3Dk(System.Security.Authentication.SslProtocols.None);k(SolutionItemsProject);k(DevLang-csharp)%26rd%3Dtrue&view=netcore-3.1)).
I hope I could answer your question.

File: tools/specs/Microsoft.Data.SqlClient.nuspec [hunk 14]
```diff
@@ -74,24 +74,24 @@ When using NuGet 3.x this package requires at least version 3.4.</description>
     </frameworkAssemblies>
     <references>
       <group targetFramework="net46">
-          <reference file="Microsoft.Data.SqlClient.dll" />
-          <reference file="Microsoft.Data.SqlClient.pdb" />
-          <reference file="Microsoft.Data.SqlClient.xml" />
+        <reference file="Microsoft.Data.SqlClient.dll" />
```
Comment (karinazhou) [inline tools/specs/Microsoft.Data.SqlClient.nuspec:77]: Revert the change in this file.

File: doc/snippets/Microsoft.Data.SqlClient/SqlClientLogger.xml [hunk 15]
```diff
@@ -37,5 +37,12 @@
             <summary>Logs information through a specified method of the current instance type.</summary>
             <remarks>To be added.</remarks>
         </LogInfo>
+        <LogWarning>
+          <param name="type">The type to be logged.</param>
+          <param name="method">The logging method.</param>
+          <param name="message">The message to be logged.</param>
+          <summary>Logs warning through a specified method of the current instance type.</summary>
+          <remarks>To be added.</remarks>
```
Comment (JRahnama) [inline doc/snippets/Microsoft.Data.SqlClient/SqlClientLogger.xml:45]: if there is nothing to be added please remove the tag.

Comment (cheenamalhotra) [inline doc/snippets/Microsoft.Data.SqlClient/SqlClientLogger.xml:45]: Avoid adding "To be added" for new documentations.. 
Either provide remarks or remove this XML element.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserHelperClasses.cs [hunk 16]
```diff
@@ -917,4 +918,59 @@ private void ParseMultipartName()
...
+        {
+            string name;
+
+            /* The SslProtocols.Tls13 is supported by netcoreapp3.1 and later
+             * This driver does not support this version yet!
+            if ((protocol & SslProtocols.Tls13) == SslProtocols.Tls13)
+            {
+                name = "TLS 1.3";
+            }*/
+            if((protocol & SslProtocols.Tls12) == SslProtocols.Tls12)
```
Comment (JRahnama) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserHelperClasses.cs:934]: A quick question what if we use switch statement instead of if and else for sslProtocols and inside each case we check the value of protocol? I think Switch is a better option for perf.

Comment (DavoudEshtehari) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserHelperClasses.cs:934]: The `switch` statement is great for fixed data values. If there are only a few cases, it might not affect the speed in any case.

File: src/Microsoft.Data.SqlClient/netcore/src/Interop/SNINativeMethodWrapper.Windows.cs [hunk 17]
```diff
@@ -225,7 +225,7 @@ internal struct SNI_Error
         internal static extern uint SNITerminate();
 
         [DllImport(SNI, CallingConvention = CallingConvention.Cdecl, EntryPoint = "SNIWaitForSSLHandshakeToCompleteWrapper")]
-        internal static extern uint SNIWaitForSSLHandshakeToComplete([In] SNIHandle pConn, int dwMilliseconds);
+        internal static extern uint SNIWaitForSSLHandshakeToComplete([In] SNIHandle pConn, int dwMilliseconds, out uint pProtocolVersion);
```
Comment (JRahnama) [inline src/Microsoft.Data.SqlClient/netcore/src/Interop/SNINativeMethodWrapper.Windows.cs:228]: is the M.D.S.SNI changed as well? a new function is added?

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObjectManaged.cs [hunk 18]
```diff
@@ -205,7 +205,11 @@ internal override uint EnableMars(ref uint info)
             return TdsEnums.SNI_ERROR;
         }
 
-        internal override uint EnableSsl(ref uint info) => SNIProxy.Singleton.EnableSsl(Handle, info);
+        internal override uint EnableSsl(ref uint info)
+        {
+            uint result = SNIProxy.Singleton.EnableSsl(Handle, info);
+            return result;
+        }
```
Comment (David-Engel) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObjectManaged.cs:212]: Seems like this should be reverted since it's functionally the same as before.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObjectNative.cs [hunk 19]
```diff
@@ -300,7 +320,8 @@ internal override uint EnableMars(ref uint info)
         internal override uint EnableSsl(ref uint info)
         {
             // Add SSL (Encryption) SNI provider.
-            return SNINativeMethodWrapper.SNIAddProvider(Handle, SNINativeMethodWrapper.ProviderEnum.SSL_PROV, ref info);
+            uint result = SNINativeMethodWrapper.SNIAddProvider(Handle, SNINativeMethodWrapper.ProviderEnum.SSL_PROV, ref info);
+            return result;
```
Comment (David-Engel) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObjectNative.cs:324]: Seems like this should be reverted since it's functionally the same as before.
